Add Array spec

LuisFSegalla opened the pull request:
Implemented Array spec based on #168. Tried keeping it similar with what was developed for **Scan Point Generator**.

Development was tested against real hardware in P99 and results seem to be what I expected.

LuisFSegalla committed:
Initial implementation of Array Spec

LuisFSegalla committed:
Fix comments

LuisFSegalla committed:
Fixed test

LuisFSegalla committed:
Added tests for coverage

LuisFSegalla committed:
Fixed Array duration method test

coretl commented on src/scanspec/specs.py:
```diff
@@ -885,6 +886,123 @@ def calculate(  # noqa: D102
+        new_dim: Dimension[Axis] = Dimension(
+            {self.axis: np.asarray(mid)},
+            {self.axis: np.asarray(lower)},
+            {self.axis: np.asarray(upper)},
```
I think there's some clever things in Dimension that avoids doing maths if the dictionaries for lower and upper are the same, can we do that for the `None` case?

LuisFSegalla replied:
Done!

LuisFSegalla committed:
PR comments

* Changed Array class attribute names
* Removed option for defining a Gap array
* Removed unnecessary bits of code
* Removed check for same size arrays in the Array spec

LuisFSegalla committed:
Changed Array Spec implementation

- Changed the implementation of the Array spec to make it more clean.
 - Only midpoints or bounds arrays are required and based on which
arrays were provided the missing ones will be calculated.
 - Fixed typos from previous commits

LuisFSegalla committed:
Updated tests to reflect class change

coretl commented on src/scanspec/specs.py:
```diff
@@ -885,6 +886,117 @@ def calculate(  # noqa: D102
+    array_duration: npt.NDArray[np.float64] | None = Field(
+        description="Array describing the duration of each point in the array",
+        default=None,
+    )
```
My idea was that you would still wrap `Array` with `Fly` if you wanted bounds calculations. So you could do `Array("x", [1, 2, 3])` to give you a step scan with bounds of the first point being [1, 1], or `Fly(Array("x", [1, 2, 3]))` with bounds of the first point being [0.5, 1.5]

LuisFSegalla replied:
My bad! I didn't realised that the `calculate` method in `Fly` already set the `bounds==True`.
I think it all makes sense now
